src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +28,239 @@
     @Autowired
     private ClinicService clinicService;
 
+    private static final int COMMAND_FIELD_NUMBER = 5;
+    private static final String COMMAND_FORMAT = "name;date(yyyy-MM-dd);type id;owner's id;command";
+    private static final String COMMAND_DELIMITER = ";";
+
+    private static final int PET_NAME_INDEX = 0;
+    private static final int DATE_INDEX = 1;
+    private static final int TYPE_INDEX = 2;
+    private static final int OWNER_ID_INDEX = 3;
+    private static final int COMMAND_TYPE_INDEX = 4;
+
+    private static final String PET_NAME_REGEX = "^[a-zA-Z]+$";
+    private static Pattern namePattern = Pattern.compile(PET_NAME_REGEX);
+
+    private static final String DATE_PATTERN = "yyyy-MM-dd";
+
+    private static Set<String> supportedCommands = new HashSet<>();
+    static {
+        supportedCommands.add("add");
+        supportedCommands.add("delete");
+    }
+
     @PreAuthorize("hasRole(@roles.OWNER_ADMIN)")
     @RequestMapping(value = "importPets",
         method = RequestMethod.POST,
         consumes = "text/plain",
         produces = "application/json")
     public ResponseEntity<List<Pet>> importPets(@RequestBody String csv) {
 
-        int i = 0;
-        List<Pet> pets = new LinkedList<Pet>();
-        Pet pet;
+        List<Pet> response = new ArrayList<>();
+        String[] commands = csv.split("\n");
+
+        for(String command: commands){
+            String[] fields = parseCommand(command);
 
-        do {
-            pet = new Pet();
+            String name = fields[PET_NAME_INDEX];
+            String dateString = fields[DATE_INDEX];
+            String typeIDString = fields[TYPE_INDEX];
+            String ownerIDString = fields[OWNER_ID_INDEX];
+            String commandType = fields[COMMAND_TYPE_INDEX].toLowerCase();
 
-            String field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
+            String errorMessage = validateCommand(fields, command);
+            if(!errorMessage.isEmpty()){
+                return createHttpError(errorMessage, HttpStatus.BAD_REQUEST);
```

Comment:
shall we skip the faulty entries and log the errors (and potentially feedback in the end), instead of terminating the whole process?

Reply:
I also thought about this for a bit, there's generally no one right way to do it, but I decided to follow the "terminal application" pattern, e.g. bash. This is less error prone because some commands can depend on the previous ones.